ethereum/eth2.0-specs
# Pythonize epoch transitions

This is first and foremost a refactor of the epoch transition to define it entirely using python code.

To make the code executable, a number of substantive changes were also included. See below for a detailed change log of such changes.

## Substantive changes 
* Merkleization friendly pending attestations (#697)
    * Motivation
        * cleaner selection of attestations from previous and current epoch
        * merkleization friendly when cleaning up attestations at end of epoch transition
    * Changelog
        * remove `latest_attestations` from `BeaconState`
        * add `previous_epoch_attestations` and `current_epoch_attestations` to `BeaconState`
        * when processing attestations from block, append to previous/current epoch attestations array depending on if `attestation.data.slot` is from previous or current epoch
        * at end of epoch transition clean up attestations via the following:
            * Set `state.previous_epoch_attestations = state.current_epoch_attestations`.
            * Set `state.current_epoch_attestations = []`.
* parent root reconstruction + granular state roots + body/header segregation (#649)
    * Motivation
        * We want state transition function signature to look like `(state, block) -> state` but there was an implicit third parameter -- `parent_root`
        * This makes a number of small changes allowing us to cache the previous block header in state to make the state "self sufficient" in the state transition
        * To this end, we added `latest_state_roots` to state which is also a nice historic feature in general.
    * Changelog
        * replace `Proposal` with `BeaconBlockHeader`. Remove `shard` field.
            * note that `hash_tree_root(block) == hash_tree_root(block_header)`
        * change `parent_root` to `previous_block_root` in `BeaconBlock`
        * move `eth1_data` and `randao_reveal` into `BeaconBlockBody`
        * add `latest_state_roots` to `BeaconState`
        * change `batched_block_roots` to `historical_roots` and use this to batch `(block, state)` rather than just `block`. Do batching at end of epoch transition.
        * add `get_temporary_block_header` helper that produces a block header from a block with a `ZERO_HASH` for state root
        * modify `get_genesis_beacon_state` to handle modified state fields
        * per-slot processing
            * Do bookkeeping related to state, block, and latest_block_header
        * per-block processing
            * add "block header section" to handle all block header verification/bookkeeping
* `get_previous_epoch` returns `current_epoch - 1` in all cases
* tie break for `winning_root` favors lexicographically _higher_ hashes rather than lower
* Inactive slashed validators only pay inactivity penalties until withdrawable (https://github.com/ethereum/eth2.0-specs/pull/711/commits/ec12460b8bae49c2272870b4490039a68f3c4761) (bugfix)
* handle `get_base_reward` when no previous balance (https://github.com/ethereum/eth2.0-specs/pull/711/commits/02e8e8974049f1965df3f7b0b9b874a46a26c150) (bugfix)
* handle case when no attestations available for a particular shard when crosslinking (https://github.com/ethereum/eth2.0-specs/pull/711/commits/47cca60148dbf74af01ef491e7b58fbac0241b75) (bugfix)
